src/meta-srv/src/region/supervisor.rs

```diff
@@ -0,0 +1,523 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt::Debug;
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+
+use common_meta::key::MAINTENANCE_KEY;
+use common_meta::kv_backend::KvBackendRef;
+use common_meta::peer::PeerLookupServiceRef;
+use common_meta::{ClusterId, DatanodeId};
+use common_runtime::JoinHandle;
+use common_telemetry::{error, info, warn};
+use common_time::util::current_time_millis;
+use error::Error::{MigrationRunning, TableRouteNotFound};
+use snafu::{OptionExt, ResultExt};
+use store_api::storage::RegionId;
+use tokio::sync::mpsc::{Receiver, Sender};
+use tokio::time::{interval, MissedTickBehavior};
+
+use super::failure_detector::RegionFailureDetector;
+use crate::error::{self, Result};
+use crate::failure_detector::PhiAccrualFailureDetectorOptions;
+use crate::handler::node_stat::Stat;
+use crate::metasrv::{SelectorContext, SelectorRef};
+use crate::procedure::region_migration::manager::RegionMigrationManagerRef;
+use crate::procedure::region_migration::RegionMigrationProcedureTask;
+use crate::selector::SelectorOptions;
+
+/// `DatanodeHeartbeat` represents the heartbeat signal sent from a datanode.
+/// It includes identifiers for the cluster and datanode, a list of regions being monitored,
+/// and a timestamp indicating when the heartbeat was sent.
+#[derive(Debug)]
+pub(crate) struct DatanodeHeartbeat {
+    cluster_id: ClusterId,
+    datanode_id: DatanodeId,
+    // TODO(weny): Considers collecting the memtable size in regions.
+    regions: Vec<RegionId>,
+    timestamp: i64,
+}
+
+impl From<&Stat> for DatanodeHeartbeat {
+    fn from(value: &Stat) -> Self {
+        DatanodeHeartbeat {
+            cluster_id: value.cluster_id,
+            datanode_id: value.id,
+            regions: value.region_stats.iter().map(|x| x.id).collect(),
+            timestamp: value.timestamp_millis,
+        }
+    }
+}
+
+pub(crate) enum Event {
```

Comment:
Can we add some doc for `Event` since those variants are crucial for understanding how to handle them in meta-srv

Reply:
@coderabbitai generates doc for `Event` struct.